lua/nvim-lsp-installer.lua

```diff
@@ -179,9 +180,11 @@ end
 
 ---@param cb fun(server: Server) @Callback to be executed whenever a server is ready to be set up.
 function M.on_server_ready(cb)
+    log.debug "Received on_server_ready"
     dispatcher.register_server_ready_callback(cb)
-    vim.schedule(function()
+    vim.schedule_wrap(function()
```

Comment:
`vim.schedule` and `vim.schedule_wrap` are the same thing, just with different semantics. Using `vim.schedule_wrap` will instead just create a function which when called will schedule the wrapped function to be executed, whereas `vim.schedule` immediately schedules the provided function to be executed in the next event loop tick. I think the reason this "fixes" the problems you're running into is because it never executes?

Reply:
I managed to make a solid reproduce in #216 for the issue. I updated the PR description with a demo for the fix as well.

Reply:
Rebased on main and `schedule_wrap` is still firing correctly, so I would still chose it over `schedule`.

Reply:
Hm interesting. It really is just a [wrapper around vim.schedule](https://github.com/neovim/neovim/blob/master/src/nvim/lua/vim.lua#L282-L287). I wonder what ends up calling it 😱

Reply:
It's not executing for me :/

This is printing on `main`, but not on this branch

```diff
diff --git a/lua/wb/lsp/init.lua b/lua/wb/lsp/init.lua
index 7d69add..7fcaa83 100644
--- a/lua/wb/lsp/init.lua
+++ b/lua/wb/lsp/init.lua
@@ -49,6 +49,7 @@ function M.setup()
     }
 
     lsp_installer.on_server_ready(function(server)
+        print("Server ready callback called " .. server.name)
         local default_opts = {
             on_attach = common_on_attach,
             capabilities = capabilities.create {
```

Reply:
Still works for me after the rebase, here's the logfile.

<details>
<summary>Click to expand</summary>

```logtalk
[DEBUG Sat 30 Oct 2021 05:02:19 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:393: Opening window
[DEBUG Sat 30 Oct 2021 05:02:20 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:199: Deleting window
[DEBUG Sat 30 Oct 2021 05:02:28 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:393: Opening window
[DEBUG Sat 30 Oct 2021 05:02:31 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:199: Deleting window
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] /home/hatsu/.config/nvim/ftplugin/java.lua:24: Automatic server installation detected
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...art/nvim-lsp-installer/lua/nvim-lsp-installer/server.lua:105: Received request to install server "jdtls".
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:612: Installing server {
  _default_options = {
    cmd = { "java", "-Declipse.application=org.eclipse.jdt.ls.core.id1", "-Dosgi.bundles.defaultStartLevel=4", "-Declipse.product=org.eclipse.jdt.ls.core.product", "-Dlog.protocol=true", "-Dlog.level=ALL", "-Xms1g", "-Xmx2G", "-jar", "/home/hatsu/.local/share/nvim/lsp_servers/jdtls/plugins/org.eclipse.equinox.launcher_*.jar", "-configuration", "/home/hatsu/.local/share/nvim/lsp_servers/jdtls/config_linux", "-data", "/home/hatsu/workspace", "--add-modules=ALL-SYSTEM", "--add-opens", "java.base/java.util=ALL-UNNAMED", "--add-opens", "java.base/java.lang=ALL-UNNAMED" }
  },
  _installer = <function 1>,
  _on_ready_handlers = {},
  homepage = "https://github.com/eclipse/eclipse.jdt.ls",
  name = "jdtls",
  on_ready = <function 2>,
  root_dir = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls",
  <metatable> = <1>{
    __index = <table 1>,
    _setup_install_context = <function 3>,
    get_default_options = <function 4>,
    get_supported_filetypes = <function 5>,
    install = <function 6>,
    install_attached = <function 7>,
    is_installed = <function 8>,
    new = <function 9>,
    on_ready = <function 10>,
    setup = <function 11>,
    uninstall = <function 12>
  }
} nil
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...cker/start/nvim-lsp-installer/lua/nvim-lsp-installer.lua:183: Received on_server_ready
[INFO  Sat 30 Oct 2021 05:02:38 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:547: Starting install server_name="jdtls", requested_version="N/A"
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:41: fs: mkdirp /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="wget", spawn_opts={
  args = { "-nv", "-O", "archive.tar.gz", "https://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:02:38 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 10610
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=10610 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="tar", spawn_opts={
  args = { "-xvf", "archive.tar.gz" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 10734
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=10734 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:22: fs: rmrf /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp/archive.tar.gz
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="wget", spawn_opts={
  args = { "-nv", "-O", "lombok.jar", "https://projectlombok.org/downloads/lombok.jar" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:02:43 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 10737
[DEBUG Sat 30 Oct 2021 05:02:44 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=10737 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:02:44 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:51: fs: mkdir /home/hatsu/.local/share/nvim/lsp_servers/jdtls
[DEBUG Sat 30 Oct 2021 05:02:44 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:33: fs: rename /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp /home/hatsu/.local/share/nvim/lsp_servers/jdtls
[INFO  Sat 30 Oct 2021 05:02:44 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:566: Installation completed server_name="jdtls", success=true
[DEBUG Sat 30 Oct 2021 05:02:44 PM CEST] /home/hatsu/.config/nvim/ftplugin/java.lua:33: Installation complete for [jdtls] server
[DEBUG Sat 30 Oct 2021 05:02:44 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/dispatcher.lua:11: running on_server_ready for ["jdtls"]

```

</details>

Reply:
It seems like the log originates from [`dispatcher.lua`](https://github.com/williamboman/nvim-lsp-installer/pull/210/files#diff-fcba1715bc0ba0f8eff00c99da14f756e80368695c140688ba30f8b167918a34R11), no? I wonder if something else is calling the dispatcher? It's probably the server instance itself, when the server is successfully installed [here](https://github.com/williamboman/nvim-lsp-installer/blob/main/lua/nvim-lsp-installer/server.lua#L158-L161).

Reply:
I should probably use the new `on_ready()` and test again.

> It seems like the log originates from [`dispatcher.lua`](https://github.com/williamboman/nvim-lsp-installer/pull/210/files#diff-fcba1715bc0ba0f8eff00c99da14f756e80368695c140688ba30f8b167918a34R11), no?

Which one? Because I added the one that says `Installation complete for [jdtls] server`, and the one above it comes from here: 
https://github.com/williamboman/nvim-lsp-installer/blob/122332d8c09071821bf0054e739c7392667dbe0e/lua/nvim-lsp-installer/ui/status-win/init.lua#L566

Reply:
Yup still works for me.

<details>
<summary>java.lua</summary>

```lua
local servers = require("nvim-lsp-installer.servers")
local log = require("nvim-lsp-installer.log")

local server_name = "jdtls"
local config = {}

local server_available, requested_server = servers.get_server(server_name)

if server_available then
	if not requested_server:is_installed() then
		log.debug("Automatic server installation detected")
		vim.notify("Automatic server installation detected", vim.log.levels.INFO)
		requested_server:install()
	end
end

requested_server:on_ready(function()
	requested_server:setup(config)
		log.debug(string.format(">>> Installation complete for [%s] server <<<", requested_server.name))
		vim.notify(string.format(">>> Installation complete for [%s] server <<<", requested_server.name), vim.log.levels.INFO)
end)
```

</details>

<details>
<summary>lsp-installer.log</summary>

```logtalk
[DEBUG Sat 30 Oct 2021 05:29:22 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:393: Opening window
[DEBUG Sat 30 Oct 2021 05:29:23 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:199: Deleting window
[DEBUG Sat 30 Oct 2021 05:29:29 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:393: Opening window
[DEBUG Sat 30 Oct 2021 05:29:29 PM CEST] ...nvim-lsp-installer/lua/nvim-lsp-installer/ui/display.lua:199: Deleting window
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] /home/hatsu/.config/nvim/ftplugin/java.lua:15: Automatic server installation detected
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] ...art/nvim-lsp-installer/lua/nvim-lsp-installer/server.lua:105: Received request to install server "jdtls".
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:612: Installing server {
  _default_options = {
    cmd = { "java", "-Declipse.application=org.eclipse.jdt.ls.core.id1", "-Dosgi.bundles.defaultStartLevel=4", "-Declipse.product=org.eclipse.jdt.ls.core.product", "-Dlog.protocol=true", "-Dlog.level=ALL", "-Xms1g", "-Xmx2G", "-jar", "/home/hatsu/.local/share/nvim/lsp_servers/jdtls/plugins/org.eclipse.equinox.launcher_*.jar", "-configuration", "/home/hatsu/.local/share/nvim/lsp_servers/jdtls/config_linux", "-data", "/home/hatsu/workspace", "--add-modules=ALL-SYSTEM", "--add-opens", "java.base/java.util=ALL-UNNAMED", "--add-opens", "java.base/java.lang=ALL-UNNAMED" }
  },
  _installer = <function 1>,
  _on_ready_handlers = {},
  homepage = "https://github.com/eclipse/eclipse.jdt.ls",
  name = "jdtls",
  root_dir = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls",
  <metatable> = <1>{
    __index = <table 1>,
    _setup_install_context = <function 2>,
    get_default_options = <function 3>,
    get_supported_filetypes = <function 4>,
    install = <function 5>,
    install_attached = <function 6>,
    is_installed = <function 7>,
    new = <function 8>,
    on_ready = <function 9>,
    setup = <function 10>,
    uninstall = <function 11>
  }
} nil
[INFO  Sat 30 Oct 2021 05:29:40 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:547: Starting install server_name="jdtls", requested_version="N/A"
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:41: fs: mkdirp /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="wget", spawn_opts={
  args = { "-nv", "-O", "archive.tar.gz", "https://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:29:40 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 20108
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=20108 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="tar", spawn_opts={
  args = { "-xvf", "archive.tar.gz" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 20253
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=20253 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:22: fs: rmrf /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp/archive.tar.gz
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:116: Spawning cmd="wget", spawn_opts={
  args = { "-nv", "-O", "lombok.jar", "https://projectlombok.org/downloads/lombok.jar" },
  cwd = "/home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp"
}
[DEBUG Sat 30 Oct 2021 05:29:46 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:162: Spawned with pid 20260
[DEBUG Sat 30 Oct 2021 05:29:47 PM CEST] ...rt/nvim-lsp-installer/lua/nvim-lsp-installer/process.lua:148: Job pid=20260 exited with exit_code=0, signal=0
[DEBUG Sat 30 Oct 2021 05:29:47 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:51: fs: mkdir /home/hatsu/.local/share/nvim/lsp_servers/jdtls
[DEBUG Sat 30 Oct 2021 05:29:47 PM CEST] ...r/start/nvim-lsp-installer/lua/nvim-lsp-installer/fs.lua:33: fs: rename /home/hatsu/.local/share/nvim/lsp_servers/jdtls.tmp /home/hatsu/.local/share/nvim/lsp_servers/jdtls
[INFO  Sat 30 Oct 2021 05:29:47 PM CEST] ...-installer/lua/nvim-lsp-installer/ui/status-win/init.lua:566: Installation completed server_name="jdtls", success=true
[DEBUG Sat 30 Oct 2021 05:29:47 PM CEST] /home/hatsu/.config/nvim/ftplugin/java.lua:24: >>> Installation complete for [jdtls] server <<<
```

</details>

Reply:
:+1: so that's working as expected when using the `server:on_ready()` API? But if you were to use the `lsp_installer.on_server_ready(callback)` one, you'd never end up executing the provided callback, when using `schedule_wrap`